feat(resources): implements `service.instance.id`

maryliag opened the pull request:
## Which problem is this PR solving?

Implements [service.instance.id](https://github.com/open-telemetry/semantic-conventions/pull/312)

## Short description of the changes

From the [Semantic Convention on Service](https://github.com/open-telemetry/semantic-conventions/tree/main/docs/resource#service), we have the values of `service.name` as required, as `service.instance.id`, `service.version` and `service.namespace` as recommended.
We are not able to set the value of version and namespace by ourselves, but following the semantic convention for [service.instance.id](https://github.com/open-telemetry/semantic-conventions/pull/312), the value can be generated using a randomUUID (we have a similar implementation of this in other SDKs, such as the Java one).
This commit adds a value for `service.instance.id` by default, which can still be overwritten by the environment variable or cloud provider.

This commit also updates some deprecated strings and add missing `sdk.shutdown` to all `sdk.start` on the sdk test file.

## How Has This Been Tested?

- [x] Added test on `sdk.test`
- [x] Tested manually

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added

trentm commented:
Thanks for this!

> I also consider adding this to the default SDK value, but that would unnecessarily force a call for a randomUUID creation on cases that are not required, so I opted for its creation only when not value is set during the environment detector checks.

I kinda think it should be on the default Resource (https://github.com/open-telemetry/opentelemetry-js/blob/c0468673cb0cfe61d12edf69745fabfdf940d7b0/packages/opentelemetry-resources/src/Resource.ts#L52-L62), but I'm curious for other's opinions.

I think it is a bit odd that one would only get `service.instance.id` if using the EnvDetector because this default service.instance.id doesn't have anything to do with `process.env.` values.

I don't think the extra `crypto.randomUUID()` call is a real perf concern. This suggests it takes on the order of 0.1 microseconds on my machine:

```
% cat timeuuid.js
const { randomUUID } = require('crypto');
const ben = require('ben');
const ms = ben(1e6, function () {
    randomUUID();
});
console.log('crypto.randomUUID(): %dms per iteration', ms);

% node timeuuid.js
crypto.randomUUID(): 0.000105ms per iteration

% node timeuuid.js
crypto.randomUUID(): 0.000095ms per iteration

% node timeuuid.js
crypto.randomUUID(): 0.00009ms per iteration
```

It would only be called once per `Resource.default()` call, which is only once in a basic usage of NodeSDK.

maryliag commented:
ah good to hear it doesn't cause such a perf impact! 
The place you suggested is exactly the place I tried first. I can add there, just a heads up that it breaks a lot of tests, so the PR will be quite big to fix all places, but hopefully it's easy to review 😄

trentm commented:
> just a heads up that it breaks a lot of tests, so the PR will be quite big

Thanks for slogging through those.

maryliag commented:
Implementation and PR description updates. I will focus on the tests now 😄

pichlermarc commented on packages/opentelemetry-resources/src/Resource.ts:
```diff
@@ -51,13 +58,14 @@ export class Resource implements IResource {
-      [SemanticResourceAttributes.TELEMETRY_SDK_VERSION]:
-        SDK_INFO[SemanticResourceAttributes.TELEMETRY_SDK_VERSION],
+      [SEMRESATTRS_SERVICE_NAME]: defaultServiceName(),
+      [SEMRESATTRS_SERVICE_INSTANCE_ID]: randomUUID(),
```
Hmm, IMO this is a quite substantial change.

I'd like to avoid adding these as they're still experimental and the `@opentelemetry/resource` package is stable. I'd prefer some form of opt-in mechanism for those.

It'll also likely affect metrics quite a bit as the resource is considered identifying for a metrics stream substantially increasing cardinality. 

We also always merge the resource with `Resource.default()` in SDKs, so there's no way to opt out from it.
We may have to change our SDKs first as it looks like we incorrectly implement this part of the spec:

> The SDK MUST provide access to a Resource with at least the attributes listed at
[Semantic Attributes with SDK-provided Default Value](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#semantic-attributes-with-sdk-provided-default-value).
This resource MUST be associated with a `TracerProvider` or `MeterProvider`
**if another resource was not explicitly specified**.

We, however, always merge `Resource.default()` with the provided one, which is incorrect:
- https://github.com/open-telemetry/opentelemetry-js/blob/47444f2a4c13a80f3328ffafc5b4c8db9e9a0886/packages/opentelemetry-sdk-trace-base/src/BasicTracerProvider.ts#L83
- https://github.com/open-telemetry/opentelemetry-js/blob/0f6518dd63bf86f3149437b2aaa56c880d45ef6d/packages/sdk-metrics/src/MeterProvider.ts#L49
- https://github.com/open-telemetry/opentelemetry-js/blob/65474404326d6ffbdd8e84799e88a46bfe66632a/experimental/packages/sdk-logs/src/LoggerProvider.ts#L37

I wonder if there's a conflicting spec somewhere of if we've just done it wrong the whole time. :fearful: 
Before we can proceed with this PR, we first need to sort this situation.

maryliag replied:
Not sure if this is incorrect implementation, because the merge function uses the non-default as precedent, and only if there isn't a value it uses what the default have, meaning this part "if another resource was not explicitly specified." is working as expected, no?

And again, for the service.instance.id the use of random is only when is not set by something else, so for example, for users who want to avoid the cardinality problem, they can set the environment variable, which will take precedent. These is also a few cloud providers who set that value, so again it would be a problem for those cases.


Considering this value is part of the recommended values, I think makes sense to implemented as part of the stable package.
WDYT?

pichlermarc replied:
> Not sure if this is incorrect implementation, because the merge function uses the non-default as precedent, and only if there isn't a value it uses what the default have, meaning this part "if another resource was not explicitly specified." is working as expected, no?

Hmm, I see. I think we may have different interpretations of that spec section. :thinking: 

**Walking through my thought process:**

> The SDK MUST provide access to a Resource with at least the attributes listed at
[Semantic Attributes with SDK-provided Default Value](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#semantic-attributes-with-sdk-provided-default-value).

From my understanding this is `Resource.default()`

> This resource MUST be associated with a TracerProvider or MeterProvider
if another resource was not explicitly specified.

I read this as:

> `Resource.default()` MUST be associated with a TracerProvider or MeterProvider if another resource (the one passed via the constructor in our case) was not explicitly specified.

Meaning that the one passed via the constructor should replace (not merge) with the default resource. I think Resource Attributes and Resource are two different things, and the way I read it it's actually Resource. I see this further solidified by [this part](https://github.com/open-telemetry/opentelemetry-specification/blob/f3511a5ccda376dfd1de76dfa086fc9b35b54757/specification/resource/sdk.md?plain=1#L38-L39):

> Note: This means that it is possible to **create** and **associate** a resource that
does not have all or any of the SDK-provided attributes present.

Meaning needs to be possible to **create** a resource that does not have `service.name`, `service.namespace`, `service.instance.id`, `service.telemetry.sdk`, ... (this is possible in our SDK)

But also meaning that users should be able to **associate** such a resource with an SDK. Which is currently not possible as we always merge the provided one with the defaults.

That prompted me to work my way through other language implementations:
- Java MeterProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java#L60) 
- Java TracerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkTracerProviderBuilder.java#L68-L72)
- Java LoggerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerProviderBuilder.java#L42-L46)
- Python MeterProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-python/blob/f56b8013f7ec5b0795f3f8945d2ebff20b15bcc3/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/__init__.py#L389-L395)
- Python TracerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-python/blob/f56b8013f7ec5b0795f3f8945d2ebff20b15bcc3/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L385-L388)
- Python LoggerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-python/blob/f56b8013f7ec5b0795f3f8945d2ebff20b15bcc3/opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/__init__.py#L604-L607)

I also looked at the C++ implementation but it's spread over 3 files for each implementation so it's harder to link compared to the ones above. It looks like the same is happening there.

Our challenge here is that we don't offer a way for users to choose if they want their `Resource` merged with the defaults, so if someone does not want the ones that are added in this PR, they will not have no choice about it. Especially with every metric producer suddenly becoming unique, this can get expensive quickly.

**A possible course of action:**

We cannot easily remove this default from the stable packages, as it would break the stability of their telemetry, possibly breaking dashboards and more in the process:

- We introduce a flag for each of the SDKs that allows users to opt-out of auto-merging with `Resource.default()`
- We introduce that same flag in NodeSDK so that users can set it there as well
- In SDK 2.0 we align with the other SDK implementations and the spec, where we only use `Resource.default()` when the user does NOT provide a Resource to the SDK constructor.

> Considering this value is part of the recommended values, I think makes sense to implemented as part of the stable package.
>
> WDYT?

I think I disagree with this. Adding experimental semconv attributes to a widely used SDK default puts us in difficult spot regarding breaking semconv changes. Especially as we've already marked the metrics/traces/resource packages as stable. Looking at what happened with [the HTTP semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/ccb6db6f3dbc6fdb4c2ec16b0ed8fccd1e783e4b/docs/http/migration-guide.md?plain=1#L3-L25), we're now required to double-emit telemetry with two different versions of semantic conventions, I'd like to avoid this here, as it's very hard to implement and maintain (we have not even figured out an elegant way to do this, see #4572). The impact would even be bigger than for the HTTP semconv as every SDK user would be affected.

If these attributes were already marked as stable, we'd just move forward with it as we don't have to fear breaking changes. Do we know what other SIGs do with these attributes? :thinking:

pyohannes replied:
@maryliag pinged me on this, I'll have a closer look at what the specification mandates here.

From a user point of view, however, it is beneficial to have both options: extend (merge with) the default resource, and also override (replace) the default resource. Blocking any of those two ways will block certain use cases.

.NET, for example, allows extending existing resources (via a [`ConfigureResource`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/OpenTelemetryBuilderSdkExtensions.cs#L36C41-L36C58) method on provider builders), as well as replacing the default resource (via a [`SetResourceBuilder`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Builder/MeterProviderBuilderSdk.cs#L139) method on the provider builder). While this can probably realized in a more elegant way, the important thing is that it covers all use cases.

pichlermarc replied:
> From a user point of view, however, it is beneficial to have both options: extend (merge with) the default resource, and also override (replace) the default resource. Blocking any of those two ways will block certain use cases.

Yep, I completely agree.

> .NET, for example, allows extending existing resources (via a [ConfigureResource](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/OpenTelemetryBuilderSdkExtensions.cs#L36C41-L36C58) method on provider builders), as well as replacing the default resource (via a [SetResourceBuilder](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Builder/MeterProviderBuilderSdk.cs#L139) method on the provider builder). While this can probably realized in a more elegant way, the important thing is that it covers all use cases.

Yes, [Java implements this in a similar way](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java#L60), `setResource()` overrides, `addResource()` merges.

Maybe this is a bigger deal in my mind than it actually is:

As @maryliag correctly pointed out earlier: users can override those attributes specifically via the env provider, or via the cloud providers (they can also do it directly in code if they'd like). 

My preference is to have feature ready so that users don't have to do it every time a new default attribute is added. The easiest way to facilitate that seems to be to implement requirement from the spec. I'll prepare a PR to show how this could look like. I'll also put it on the agenda for tomorrow's SIG meeting. I'm sure we can figure something out.

maryliag replied:
Those next steps make sense to me, thank you both for the input!

maryliag replied:
Regarding the topic of this value being always set VS with opt-in: I used the example of how other stable SDKs deal with that (e.g. [here](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore#experimental-support-for-grpc-requests)), and I made a change in this PR to have the value as default as opt-in based on the env var `OTEL_NODE_JS_EXPERIMENTAL_DEFAULT_SERVICE_INSTANCE_ID`

This way this feature can still be part of the stable package, but with experimental values. And after we have that for awhile, we can change the behaviour in the future.

Then we can have another PR focusing on the merge/override, like the one @pichlermarc created on #4617

maryliag commented on packages/opentelemetry-resources/src/Resource.ts:
```diff
@@ -50,15 +58,22 @@ export class Resource implements IResource {
+    if (getEnv().OTEL_NODE_JS_EXPERIMENTAL_DEFAULT_SERVICE_INSTANCE_ID) {
+      defaultResource.attributes[SEMRESATTRS_SERVICE_INSTANCE_ID] =
+        randomUUID();
+    }
```
interesting! Changes made.

pichlermarc commented:
Short summary of what we discussed in the SIG Meeting today :slightly_smiling_face: 
- we'll move the detection of this attribute to it's own resource detector in the `@opentelemetry/resources` package.
- people that currently set up their SDKs manually can use that new resource detector in the setup code (this is their opt-in)
- we add the detector in `NodeSDK` if an env var is set AND no list of resource detectors is passed via the `resourceDetectors` constructor option
  - we can use `OTEL_NODE_RESOURCE_DETECTORS` for this (which contains a comma seperated list of resource detectors to turn on), the values for existing resource detectors should match [this list](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/1af49014b8a77d34058750ce2a87bd58211ebe70/metapackages/auto-instrumentations-node/src/utils.ts#L77-L84), we'll discuss the value for the new detector on the PR - `instance` or `serviceinstance` comes to mind.

Implementing the resource detector and adding the env var behavior in NodeSDK can be done in seperate PRs to speed up the review process. :rocket: 

Separately from that:
- I'll implement a way for users to opt-out of auto-merging with the default resource in MeterProvider and TracerProvider akin to #4617 
- I'll also implement the change in behavior on the `next` branch (SDK 2.0) which will then be fully spec compliant 
  - (I actually committed the metrics part of the crime against the spec in #3411 :cry: Sorry about that :slightly_frowning_face: )

maryliag commented:
Latest changes (and some questions):

1. Created `serviceInstanceIDDetector` 
2. The Browser version does nothing and the Node version sets the value with `randomUUID`
3. The default resource is working the same way. I removed the changes I had made there (just keep the updates on deprecated values)
4. I added the new detector on the default list of detectors only if the env var `OTEL_NODE_EXPERIMENTAL_DEFAULT_SERVICE_INSTANCE_ID` is set. The other condition will come on a later PR since changes are required on the js-contrib repo first, then I can come back to this one

Looking for some help: I added a few tests for a few cases with value coming fro env var VS should get set based on the experimental flag and one of the tests is failing but I don't quite get why.
The test failing is where I don't pass a service.instance.id on env var, but I do set to true the `OTEL_NODE_EXPERIMENTAL_DEFAULT_SERVICE_INSTANCE_ID`, so I should see a random UUID for its value, but on the test itself, it's returning `undefined`. When I print [here](https://github.com/open-telemetry/opentelemetry-js/pull/4608/files#diff-f84b79e46ee7f0118d6da022e9db89e854679146fb85c5ea87fa73bc1b80028bR706) the value of resources, I don't see anything getting set for `service.instance.id`. This problem is not particular to the implementation of using the detectors, because I was having the same error when setting the value on the `defaultResources`

maryliag commented:
while working on the following PR I realize it makes more sense to have all the logic of the env var on the other one, so I remove those from here. This PR focus on the creation of the service instance ID detector. It currently is not called anywhere, but we have a test showcasing that it works as expected if the user initializes their NodeSDK with it.

The following PR will add the `OTEL_NODE_EXPERIMENTAL_DEFAULT_SERVICE_INSTANCE_ID` and `OTEL_NODE_RESOURCE_DETECTORS` with their respective logics

pichlermarc commented on packages/opentelemetry-resources/src/platform/node/ServiceInstanceIDDetector.ts:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Detector } from '../../types';
+import { ResourceDetectionConfig } from '../../config';
+import { IResource } from '../../IResource';
+import { serviceInstanceIDDetectorSync } from '../browser/ServiceInstanceIDDetectorSync';
+
+/**
+ * ServiceInstanceIDDetector detects the resources related to the service instance ID.
+ */
+class ServiceInstanceIDDetector implements Detector {
+  detect(_config?: ResourceDetectionConfig): Promise<IResource> {
+    return Promise.resolve(serviceInstanceIDDetectorSync.detect());
+  }
+}
+
+export const serviceInstanceIDDetector = new ServiceInstanceIDDetector();
```
We can remove this implmenetation and just implement `DetectorSync`. 

Some back-story on it:

We've deprecated `Detector` in favor of `DetectorSync` as people had to `await` resource detection even though most resources resolved synchronously. This was originally done for resources which have to be retrieved via an http call. To avoid users having to deal with `await`ing resources we introduced the concept of async-resolvable resource attributes (`IResource#waitForResourceAttributes()`) and moved the `await`ing of these to the SDKs export pipelines instead.

We kept the `Detector` implementations around because we could not break the interface. For new detectors we now recommend to only use `DetectorSync`

pichlermarc commented on packages/opentelemetry-resources/src/platform/node/ServiceInstanceIDDetectorSync.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+}
+
+export const serviceInstanceIDDetectorSync =
+  new ServiceInstanceIDDetectorSync();
```
We should put `@experimental` on those to let users know that this can change (semantic conventions for this are not yet stable, but `@opentelemetry/resources` is at `1.23.0` already) :slightly_smiling_face: 

```suggestion
/**
* @experimental
*/
export const serviceInstanceIDDetectorSync =
  new ServiceInstanceIDDetectorSync();
```

pichlermarc commented on packages/opentelemetry-resources/src/platform/browser/ServiceInstanceIDDetector.ts:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { noopDetector } from '../../detectors/NoopDetector';
+
+export const serviceInstanceIDDetector = noopDetector;
```
We can also drop this any only use the sync one. :slightly_smiling_face: 
(see longer comment on the Node.js implementation of it)

maryliag replied:
Done

pichlermarc commented on packages/opentelemetry-resources/src/platform/browser/ServiceInstanceIDDetectorSync.ts:
```diff
@@ -0,0 +1,19 @@
+
+import { noopDetectorSync } from '../../detectors/NoopDetectorSync';
+
+export const serviceInstanceIDDetectorSync = noopDetectorSync;
```
```suggestion
/**
* @experimental
*/
export const serviceInstanceIDDetectorSync = noopDetectorSync;
```

pichlermarc commented on CHANGELOG.md:
```diff
@@ -11,7 +11,8 @@ For experimental package changes, see the [experimental CHANGELOG](experimental/
 
-feat(sdk-trace-base): log resource attributes in ConsoleSpanExporter [#4605](https://github.com/open-telemetry/opentelemetry-js/pull/4605) @pichlermarc
+* feat(sdk-trace-base): log resource attributes in ConsoleSpanExporter [#4605](https://github.com/open-telemetry/opentelemetry-js/pull/4605) @pichlermarc
+* feat(resources): new detector ServiceInstanceIDDetector that sets the value for `service.instance.id` as random UUID.
```
```suggestion
* feat(resources): new experimental detector ServiceInstanceIDDetector that sets the value for `service.instance.id` as random UUID.
```

pichlermarc reviewed (commented):
Generally looks good :slightly_smiling_face: 

We can drop the `Detector` and use `DetectorSync` only. Other than that, only comments about marking things as `@experimental`.

Thanks for sticking with us through the review-process :slightly_smiling_face:

maryliag commented:
Great, thank you for all the context and the reviews! 😄

pichlermarc reviewed (approved):
Looks good :+1: 
Thank you for working on this @maryliag :slightly_smiling_face:

pichlermarc reviewed (commented):
I'll wait a bit with merging to give @trentm another chance to take a look as he initially reviewed the PR. :slightly_smiling_face: 

(Approvers: feel free to rebase/merge if you come across this and I have not done it yet :slightly_smiling_face: - I'm OOO until Monday)

trentm commented:
FYI: A currently open otel spec issue (by some OTel Java implementors) asking about whether the SDK should handle `service.instance.id` and veering into whether `service.instance.id` should vary between process restarts: https://github.com/open-telemetry/opentelemetry-specification/issues/3136

trentm reviewed (approved):
LGTM. Dealer's choice in my OCD naming nit.

maryliag commented on experimental/packages/opentelemetry-sdk-node/test/sdk.test.ts:
```diff
@@ -59,6 +59,7 @@ import {
   processDetector,
   hostDetector,
   Resource,
+  serviceInstanceIDDetectorSync,
```
OCD crisis adverted! 😄 Changed to `Id`
